### PR TITLE
feat: Syncing chunks from archival nodes

### PR DESCRIPTION
**Pull request by SkidanovAlex**
Currently we always request chunks parts either from a node to which the chunk part is assigned to, or from the chunk producer / full node tracking shard (if the chunk is assigned to us).

This works well with state sync, because statesync brings the node to a state as of epoch-two ago, and the block producers are all still online, but doesn't work well with the block sync from genesis, because for old blocks the block producers are almost certainly offline.

In this PR I make the following change:
1. If the chunk is from the current or the previous epoch, follow the logic we have today;
2. Otherwise, sync it from a peer that runs an archival node tracking the shard.

The change requires a change to the `Handshake`, so that archival nodes can report that they are archival.

Several seemingly unrelated changes that were necessary:
1. I changed the `num_data_parts` to be 1/3 of the total parts rounded up (instead of rounded down). In other words, it's `f+1` instead of `f` if `n=3f+1` is the total. Since the number of adversaries we assume is `f` or less, and the number of signatures on the block is `2f+1`, having `f+1` data parts doesn't break the data availability. Importantly, for four nodes it makes number of data parts 2, not 1, and thus the tests (including the tests for this change) actually trigger fetching chunk parts.

2. I'm changing the test infra to start proxy processes after the nodes are started, serially. Without that `Popen` sometimes hangs if the proxies for some processes happen to start before some other processes.

Test plan
---------

Adding two tests. `partial_chunk_messages` is not specific to this change, and makes sure that we do not send more requests (or messages) for chunks than we should. It's to prevent us accidentally introducing extra rounds of requests.

`sync_chunks_from_archival` tests this particular change, and makes sure that when an archival node starts, it sync old chunks from archival nodes, and the new chunks from the current block producers.

### CODE REVIEW

**Review by bowenwang1996 — COMMENTED** [review]
Why couldn't we pass `ChainStore` to `request_chunks` and avoid adding a new method to `RuntimeAdapter`?

**Comment by frol** [comment]
Can we request the blocks from whoever has them (like peers in bittorrent) instead of relying on some specific beefy nodes which may maliciously misreport that they are archival?

**Comment by SkidanovAlex** [comment]
> Can we request the blocks from whoever has them (like peers in bittorrent) instead of relying on some specific beefy nodes which may maliciously misreport that they are archival?

Nobody in your network will have such blocks and chunks except for archival nodes that track shard. Unless I don't understand the suggestion.

We will ultimately add proper protection against nodes misreporting their `archival` status. It is low-ish priority, because one can manually ban the nodes that slow down the sync and restart (syncing an archival node is such an long lasting process that such manual interventions don't add much pain relative to the other pains incurred already :)).

**Comment by SkidanovAlex** [comment]
> Why couldn't we pass ChainStore to request_chunks and avoid adding a new method to RuntimeAdapter?

Because that method uses epoch manager, not chain store?

**Comment by bowenwang1996** [comment]
> Because that method uses epoch manager, not chain store?

It seems to me that you only use epoch manager to get the epoch id and next epoch id of the header head, which I thought `Tip` already contains.

**Comment by SkidanovAlex** [comment]
> > Because that method uses epoch manager, not chain store?
> 
> It seems to me that you only use epoch manager to get the epoch id and next epoch id of the header head, which I thought `Tip` already contains.

I also use it to get the epoch and the next epoch for the chunk, but I guess that I can do with the ChainStore.
Then it would be doable, but there's another caveat: for the `KeyValueRuntime` I actually want it to not return `true`, because `cross_shard_tx` will break -- it routinely has forks more than two epochs deep.

**Comment by frol** [comment]
> Nobody in your network will have such blocks and chunks except for archival nodes that track shard. Unless I don't understand the suggestion.

If there are some nodes that don't garbage collect for 100 epochs (it is configurable, right?) and others keep only 5, you don't need to query archival nodes for those epochs that are more recent than the 100 epochs.

It might be overengineering, so feel free to disregard it, but it seems to be that archival nodes get under a heavy load since nobody else can help them with the task.

Will a half-synced archival node be able to stream its synced blocks to the newly setup archival nodes? (it is currently estimated to take 1/3 of the network lifetime to get fully synced, e.g. 3-month-old network will require a new archival node to sync for a whole month, which is a whole month of a load to a fully synced archival node)

File: core/primitives/src/version.rs [hunk 1]
```diff
@@ -18,12 +18,12 @@ pub const DB_VERSION: DbVersion = 8;
 pub type ProtocolVersion = u32;
 
 /// Current latest version of the protocol.
-pub const PROTOCOL_VERSION: ProtocolVersion = 36;
+pub const PROTOCOL_VERSION: ProtocolVersion = 37;
 /// TODO: Remove in next release. This is to allow nodes with initial version 34
 /// to be compatible with nodes at version 35
 pub const NETWORK_PROTOCOL_VERSION: ProtocolVersion = 34;
```
[the remark this answers is not in the document]
Comment (SkidanovAlex) [reply core/primitives/src/version.rs:24]: I didn't add this in this PR, as seen from it being white, not green :)

[the remark this answers is not in the document]
Comment (mfornet) [reply core/primitives/src/version.rs:24]: yep, but I removed it in the branch you rebase into AFAIR

[the remark this answers is not in the document]
Comment (mfornet) [reply core/primitives/src/version.rs:24]: It seems I didn't remove it! This is safe (and expected) to remove now.

**Comment by bowenwang1996** [comment]
> I also use it to get the epoch and the next epoch for the chunk, but I guess that I can do with the ChainStore.

For that you can just use runtime adapter, which I assume in `KeyValueRuntime` will return what you want.

File: chain/network/src/peer.rs [hunk 2]
```diff
@@ -273,16 +273,32 @@ impl Peer {
...
-                        act.protocol_version,
-                        act.node_id(),
-                        act.peer_id().unwrap(),
-                        act.node_info.addr_port(),
-                        PeerChainInfo { genesis_id, height, tracked_shards },
-                        act.edge_info.as_ref().unwrap().clone(),
-                    );
-                    act.send_message(PeerMessage::HandshakeV2(handshake));
+                    let handshake = match act.protocol_version {
+                        36..=37 => PeerMessage::Handshake(Handshake::new(
```
Comment (mfornet) [inline chain/network/src/peer.rs:279]: It is better to use `36..=PROTOCOL_VERSION` in case `PROTOCOL_VERSION` needs to be increased.

**Comment by SkidanovAlex** [comment]
> For that you can just use runtime adapter, which I assume in KeyValueRuntime will return what you want

In KeyValueRuntime  I want completely different behavior: I don't want it to ever sync from archival nodes. The tests that use it do not expect this logic.

**Review by bowenwang1996 — APPROVED** [review]

**Comment by pczarn** [comment]
@SkidanovAlex :
> I'm changing the test infra to start proxy processes after the nodes are started, serially. Without that Popen sometimes hangs if the proxies for some processes happen to start before some other processes.

How can I reproduce the problem with Popen?

I noticed that proxy-related tests such as `sanity/proxy_{simple,example}` are failing. After some digging, I found that the mentioned change is the cause. It appears that proxies do not spin up fast enough to accept inter-node connections before they are made. I will be looking for solutions to the issue.

I need proxies to be functional as a part of #31 and #3661

**Pull request merged**